Move to es6

albertjan opened the pull request:
I understand this is a very big PR and will make merging with other PR difficult. I don't mind if those get merged first and that I have to rebase this. 

First and foremost this is a translation to es6, all the `.bind(this)`'s in the code made me so confused :smile:. es introduces the fat arrow method `=>` that will keep the same scope as the function it's in. It also introduces string literals. :tada:! Something very similar to the `'''` strings in python.

You have to transpile the src/pypyjs.js to lib/pypyjs.js by running the gulp script in the root. You'll need to run `npm install` in the root and `npm install gulp -g`. 

Gulp also lints the code using eslint. the configuration for that is in `.eslintrc`

This also fixes a bug in the repl because it didn't use the new `top_level_scope`.

kirbyfan64 commented:
What browsers does this work on? Some of them don't support all of ES6 yet.

albertjan commented:
@kirbyfan64 it's translate to es5 with babel so browser support doesn't change. :smile:

crantila commented:
Could you please add instructions to the README? I understand the arguments in favour of ES6, but that's a matter of principle&mdash;my level of comfort with JavaScript is still such that anything starting with "just run ``npm install`` in the root directory" inevitably feels like "just run ``npm install`` then do fffffffff zzzzzz rrrrrrr until it works but probably not in the intended way." And I don't know how to run gulpfiles, or how to predict what will happen when I do.

I know these are common tools, and it's basic info, but a little bit of explanation can be *very* helpful for those of us who are still learning! Thanks!

(Actually, similar request with ``addModule()``. Most of the PyPy.js functions have at least a one-line explanation in ``pypyjs.js``).

albertjan commented:
@crantila will do! :+1:

albertjan commented:
@crantila added some docs

crantila commented:
Thanks, that's just what I was hoping for!

rfk commented:
Sounds great!  This is indeed a big PR, taking a look now...

albertjan commented:
@rfk I'm on hand atm, if you have any questions fire away :smile:

rfk commented on README.rst:
```diff
@@ -49,3 +49,21 @@ Repository Overview
+Transpiling es6
+~~~~~~~~~~~~~~~
+
+Pypyjs is written using es6 and is transpiled to es5 using `babeljs`_ to run the
```
We've been trying to use "PyPy.js" consistently when referring to the project; otherwise everyone makes up their own unique capitalization ;-)

rfk replied:
Docs nit: "to run" feels like is should start a new sentence, i.e. "...using babeljs.  To run the..."

albertjan replied:
yessir :smile: pYpYjs it is

rfk commented on README.rst:
```diff
@@ -49,3 +49,21 @@ Repository Overview
+`npm install gulp -g`.
+
+When you run `gulp` in the repository the es6 source in /src is transpiled to
+es5 and thats copied to /lib.
```
Should we also add this to the various recipes in the Makefile, to ensure it gets run when building releases etc?

albertjan replied:
I'll dive into that I don't have much experience with makefiles.

rfk replied:
I'm also happy to take a look as a follow-up to this PR

rfk commented on lib/tests/test.py:
```diff
@@ -0,0 +1,2 @@
+import time
+assert time.time() > 0
```
Please give this file a more descriptive name, and maybe even a short comment to explain why it exists.

rfk commented:
(by the way, I haven't used es6 before but I'm pretty excited for an opportunity to try it out!)

rfk commented on src/pypyjs.js:
```diff
@@ -0,0 +1,1103 @@
+  return this.fetch(file).then((data) => this.addModule(name, data.responseText));
+};
+
+// add a Module to the vm from javascript. It will automatically write it it to
```
"it it"

albertjan commented:
es6 is pretty sweet if you ask me :smile:

rfk commented on src/pypyjs.js:
```diff
@@ -0,0 +1,1103 @@
+    // Now we can execute the code in custom top-level scope.
+    const _code = (options && options.file)
+      ? `top_level_scope['__file__'] = '${options.file}'; execfile('${options.file}', top_level_scope.__dict__)`
+      : `exec '''${_escape(code)}''' in top_level_scope.__dict__`;
```
If you pass in `options.file`, does that mean the `code` argument is ignored?

More broadly, it's not obvious to me why `pypyjs.prototype.execfile` was refactored to pass in additional options to this method, can you say a little about what's behind that change?

albertjan replied:
Correct and correct, I will refactor/explain why this change is in there.

rfk commented on src/pypyjs.js:
```diff
@@ -0,0 +1,1103 @@
+  return new Promise(function promise(resolve, reject) {
+    const _code = `try:
+  ${_blockIndent(_preCode, '  ')}
+  ${code}
```
What's the difference between `code` and `preCode` here, and why would I put something in `preCode` rather than `code`?

albertjan replied:
That was added if you want to run something outside of the `top_level_scope`. As with the other question, I'll explain it better. Maybe it should be moved to a different function.

rfk replied:
Ah, gotcha.  It might be clearer as a separate function, or an explicit "don't use top-level scope" option.  Does ES6 offer anything fun like named function arguments?  If so we could do something like:

```
_execute_source(preCode, unscoped=true).then(function () {
  return _execute_source(code)
})
```

Rather than combining the two into a single code string.

albertjan replied:
there was no scoping going on in `_execute_source` so you were right we can just call it twice

rfk commented:
> Then there's a function that lets you reinitialise the global scope without 
> rebooting the vm. reInit() -> Promise

This seems like a risky endeavour since it could be easy to forget about some pieces of global state.  What's your use-case for this, and how much better/faster is it than just creating a `new pypyjs()` to get a fresh VM?

rfk commented on src/pypyjs.js:
```diff
@@ -0,0 +1,1103 @@
+  return this._ready
+    .then(() => this.loadModuleData('code'))
+    .then(() => this._execute_source('import code'))
+    .then(() => this._execute_source('c = code.InteractiveConsole(top_level_scope.__dict__)'))
```
nice catch re: the use of `top_level_scope` here

rfk commented:
> es6 is pretty sweet if you ask me 

First impression: I like the fat arrows a *lot*.

The string literal syntax, I suspect will take some getting used to...but it's obviously a huge win for readability of the python-code-inside-javascript-code strings we're doing here.

rfk commented on src/pypyjs.js:
```diff
@@ -0,0 +1,1103 @@
+      imports
+    };
+    if (!this.inJsModules) {
+      this.inJsModules = [];
```
Should this be a `{}` rather than `[]`?  It looks like you use string keys with it below.

albertjan replied:
Good catch :smile: it actually just works because it just adds properties to the array. But the array stays empty.

albertjan commented:
> This seems like a risky endeavour since it could be easy to forget about some pieces of global state. 

Very true, this maybe a case of optimization based on gut feeling :smile:. I try to do some tests.

rfk commented:
> addModule(name, code) -> Promise
> addModuleFromFile(name, filename) -> Promise

I'm not sure about these two, I wonder if they're operating at the wrong level of abstraction.  Their implementation seems to sit at a level above the virtual filesystem.  If you had an API for directly manipulating the VM's filesystem, would you still need these, or could you build these are wrappers on top of it?

rfk commented:
> this maybe a case of optimization based on gut feeling

Heh, FWIW I find it *very* plausible that it'll be faster, but I'd be curious just how much.

albertjan commented:
> I'm not sure about these two.

This was the shortest path to success for me there was no way to add files/modules to the vm. I agree it might be better to change the implementation to write directly to the FS. Do you think the api is wrong? It might be an idea to add `addModuleFromFile(file) -> Promise` that defaults to the name of the file for the module name.

rfk commented:
@albertjan I like where this PR is headed, but it's a lot of code to take in all at once; I'll come back to it again tomorrow for another look

albertjan commented:
@rfk absolutely! I'll try and address the comments this evening.

rfk commented:
> Do you think the api is wrong?

The API seems fine, but I think we can come up with a better abstraction to make the implementation clearer and cleaner.  (Which I guess we can always do as a follow-up PR, if the API and its semantics make sense)

albertjan commented on src/pypyjs.js:
```diff
@@ -0,0 +1,1113 @@
+        Module._pypy_setup_home(pypy_home, 0);
+        Module._free(pypy_home);
+
+        const initCode = [
```
@rfk any reason why you execute this line by line?

rfk replied:
sorry, I didn't notice this comment - no, no particular reason, other than the unreadability of multi-line strings in es5

rfk commented on src/pypyjs.js:
```diff
@@ -0,0 +1,1112 @@
+    Module.TOTAL_MEMORY = this.totalMemory;
+    Module.resolve = () => {
+      console.log('resolved without subscription');
+    };
```
What is `Module.resolve` used for?  Is this just a debugging aid?

albertjan replied:
Yes that should be removed.

rfk commented on src/pypyjs.js:
````diff
@@ -0,0 +1,1112 @@
+      //   import sys
+      //   if 'foo' in sys.modules: del(sys.modules['foo'])
+      // ```
+      const modulesToLoad =
````
nit: this feels like it should be named something like `modulesToReload` rather than `modulesToLoad`, since the constructed code doesn't actually load them.

rfk commented on src/pypyjs.js:
````diff
@@ -0,0 +1,1112 @@
+      // ```
+      const modulesToLoad =
+        Object.keys(this._modulesToReset)
+          .map(mod => `if '${mod}' in sys.modules: del(sys.modules['${mod}'])`);
````
I'd like to hear more about why this re-setting is necessary.  Is it to allow you to provide updated (e.g. live-edited) content for modules after you've already run the code once  If so, I wonder if it should live closer to the point of use rather than here in the core `exec` function, for example by directly executing this `del sys.modules` when you call `addModule` rather than delaying it until a subsequent call to `exec`.

albertjan replied:
Yes it is for live editing, you can change the code in the module "on disk" but if you use it a second time it won't be loaded from disk. 

I think you were right about the added intricacies with the `reInit` function that's the cause of this.

It might be better to add it to the `addModule` function. That would save us from having to save state between the moment we call `addModule` and `exec`. But it might be better to gut it all and just reInit a whole new vm thus solving the whole saved state stuff.

rfk replied:
It might also work as a `resetModule` method that the calling code can use when it knows it needs to reset the state, if this is an operation that needs to be done frequently.

In general though, importing modules can have side-effects that are not undone by `del sys.modules`, so re-creating a fresh VM is likely to be more reliable.

rfk commented on src/pypyjs.js:
```diff
@@ -0,0 +1,1112 @@
+      this.inJsModules = {};
+    }
+    this.inJsModules[`modules/${name}.py`] = source;
+  });
```
I've been thinking about this `inJsModules` stuff, and I *think* I have my head around it.  Here's a proposed refactor that IMHO works a little more with-the-grain of the existing loading and filesystem machinery.  It may not do exactly what you want though, so interested in your thoughts:

* make `addModule` just write the module data into the appropriate location in the VM's virtual filesystem, either directly using `_writeModuleFile` or some generalization of it.
* make `fetch` look in the virtual filesystem first, only fetching from the server if the file is not found locally

I *think* this would mean that we don't have to maintain an explicit `inJsModules` hash, but can just store state directly in the virtual filesystem.  There may be some subtleties with path-handling though.  Thoughts?

albertjan replied:
Like I said before this was the shortest path to success for me, discussing this PR with you I've thought about it more then when I was implementing it :smile:. I agree with you that using the virtual FS as a single source of truth instead of adding this new concept of inJsModules is probably a better idea.

I have to think about where to put this though because adding the check if it's on "disk" already in the `fetch` method seems a bit wrong to me. As all it does now is get things and it seem logical to keep it that way. :smile:

rfk commented:
@albertjan also, if you feel like getting the es6 stuff landed, I'd be happy to merge a PR containing just that cleanup and keep discussing the `addModule` stuff in isolation.

albertjan commented:
That may be a good idea. I hadn't expected there be this may ifs and buts with the `addModule` stuff :smile:.

albertjan commented:
right tests pass again

rfk commented:
Whoops, sorry, I didn't catch that you'd pushed new commits here!  Taking a look now.

albertjan commented:
no worries :smile: I'll rebase this in the meantime.

rfk commented on src/pypyjs.js:
```diff
@@ -0,0 +1,1013 @@
+// Expose the main pypyjs function at global scope for this file,
+// as well as in any module exports or 'window' object we can find.
+
+// Generic debugging printf.
```
Looks like this comment is leftover after removing the definition of `debug`

rfk commented on src/pypyjs.js:
```diff
@@ -0,0 +1,1013 @@
+  this._pendingModules = {};
+  this._loadedModules = {};
+  this._allModules = {};
+  this._modulesToReset = {};
```
`_modulesToReset` is no longer used in this PR

rfk commented on README.rst:
```diff
@@ -49,3 +49,21 @@ Repository Overview
 .. _pypyjs.org: https://pypyjs.org
+
+Transpiling es6
+~~~~~~~~~~~~~~~
```
Can you please move this description into CONTRIBUTING.rst?  I think it'll be a better fit there than in the main README.

rfk commented:
Otherwise, this LGTM!

> I'll rebase this in the meantime.

Great, thanks.  There are a lot of commits here with some backing out work introduced by others, is it worth squashing them into a single commit for the final merge?

albertjan commented:
We'd loose a lot of commit history if I squash them all into one. Its your call I do understand we don't want to muddy the waters with commits backed out later.

Otherwise I've addressed all your comments. Running some tests now. Will have the final green light in about an hour.

albertjan commented:
the module bundler doesn't work with py2 anymore :smile:

albertjan commented:
there's an issue with the start path on node that I'll have to look into.

rfk commented:
> We'd loose a lot of commit history if I squash them all into one

No worries, I'm happy to keep it.

rfk commented:
> the module bundler doesn't work with py2 anymore

Whoops, fixed in https://github.com/pypyjs/pypyjs/commit/e080e47a96b10d67969b32bd1143afcf622d6ad3

albertjan commented:
@rfk I got the repl to work again, but the tests have stopped working. On chrome it just kills the tab... very ominous. And on node I get a syntax error in a builtin module. _abcol.py. 

I built pypyjs.vm.js by running `make` without any args.

Could you maybe have a look around see if I did anything stupid?

rfk commented:
yikes; I'll take a look with my local build

rfk commented on src/pypyjs.js:
```diff
@@ -0,0 +1,1019 @@
+      // Now we can execute the code in custom top-level scope.
+      _code = `top_level_scope['__file__'] = '${options.file}'; execfile('${options.file}', top_level_scope.__dict__)`;
+    } else {
+      _code = `exec('''${_escape(code)}''' in top_level_scope.__dict__)`;
```
This needs to become `exec(code, dict)` not `exec(code in dict)` type throws a `TypeError`.  I have it fixed in my local copy, if that's all it takes to get the tests running then I'll just merge with the fix.

albertjan replied:
ouch ... sorry yeah good one :+1: will update asap

albertjan replied:
do you have any tips for debugging pypyjs?

rfk replied:
not really; I tend to fall back very quickly to "`console.log` all the things"

albertjan replied:
yeah me too :) thanks for looking into this! I struggled with this yesterday. It would have taken me a looong time to recognize this error.

rfk commented:
> On chrome it just kills the tab... very ominous

I'm seeing this behaviour too.  Interestingly, it doesn't crash using the non-es6 version, even with the exact some underlying `pypyjs.vm.js` file.

rfk commented:
I went ahead and merged this, with the `exec in` fix and a couple of tweaks to the makefile to run gulp automatically.  Thanks so much @albertjan for seeing this through!  We'll still have to follow up on the chrome crashing bug, but I can only imagine that being a chrome bug we'll have to work around, so we can do it in a separate issue.

albertjan commented:
re: Chrome crash, I'll look into that tonight! It's good to know it's due to something in the es6 branch, that means it isn't somewhere in asmjs. :smile:

Thanks for merging this, and fixing the make file (I should learn how to do make files)! I have more goodies on the way based on this!
